mpact/participants.py

```diff
@@ -0,0 +1,35 @@
+from collections import namedtuple
+
+from mpact.exceptions import TelegramIdNotFound
+from mpact.models import IndividualChat
+from mpact.services import get_telegram_id
+
+ParticipantInfo = namedtuple('ParticipantInfo', ['study_id', 'phone_number'])
+ParticipantImportResult = namedtuple('ParticipantImportResult', ['participant_info', 'successful', 'message'])
+
+
+async def import_participants(participants):
+    results = []
+    for participant in participants:
+        try:
+            telegram_id = await get_telegram_id(participant.phone_number)
+            try:
+                individual = IndividualChat.objects.get(id=telegram_id)
+                individual.study_id = participant.study_id
+                individual.save()
+                results.append(
+                    ParticipantImportResult(participant, True, '')
+                )
+            except IndividualChat.DoesNotExist:
+                results.append(
+                    ParticipantImportResult(participant, False, f'No participant with telegram ID {telegram_id} found')
+                )
```

Comment:
I wonder if this is the correct functionality. I believe this means that if I:

1. Create a group with 10 participants
2. Configure scheduling for the group
3. Upload study IDs for each of the 9 participants (10, minus the bot)

It will fail, correct? I would need to wait until each user individually messages the bot before it would create an `IndividualChat` and be able to upload?

I am thinking we would want to create an `IndividualChat` object for participants that don't yet exist. And perhaps we need a flag to alert the frontend that it is a read-only chat until we receive a message from the user that first time?

Let me know if I should elaborate

Reply:
Yeah, that sounds right (it would fail to import the ones who didn't have conversations, but would message the ones who didn't).

I think the fundamental issue is with the data model. E.g. we don't have a model to represent a person or study participant, we just have this hybrid thing that represents both that and the telegram chat with them. I played with pulling that out and decided it wasn't worth it, but guessing it is due to situations like this.